From def30e6bb8b1b16d8ac2d26b41af1dfdb405bc6c Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Fri, 16 May 2014 23:52:45 +1000
Subject: [PATCH 2/3] Cache for 500ms the result of
 RemoteEncoder::GetFramesWritten()

This function queries the backend, and when seeking several calls could be made to it which blocks the GUI thread. Cache the value for 500ms in order to alleviate the issue
---
 mythtv/libs/libmythtv/remoteencoder.cpp |   16 +++++++++++++---
 mythtv/libs/libmythtv/remoteencoder.h   |    2 ++
 2 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/mythtv/libs/libmythtv/remoteencoder.cpp b/mythtv/libs/libmythtv/remoteencoder.cpp
index ed009fc..8a335ad 100644
--- a/mythtv/libs/libmythtv/remoteencoder.cpp
+++ b/mythtv/libs/libmythtv/remoteencoder.cpp
@@ -16,6 +16,8 @@ using namespace std;
 
 #define LOC QString("RemoteEncoder(%1): ").arg(recordernum)
 
+#define MAX_SIZE_CHECK 500  // in ms
+
 RemoteEncoder::RemoteEncoder(int num, const QString &host, short port)
     : recordernum(num),       controlSock(NULL),      remotehost(host),
       remoteport(port),       lastchannel(""),        lastinput(""),
@@ -196,20 +198,28 @@ float RemoteEncoder::GetFrameRate(void)
  *         instance.
  *
  *  \sa TVRec::GetFramesWritten(void), EncoderLink::GetFramesWritten(void)
- *  \return Number of frames if query succeeds, -1 otherwise.
+ *  \return Number of frames if query succeeds, return last known value otherwise.
  */
 long long RemoteEncoder::GetFramesWritten(void)
 {
+    if (lastTimeCheck.isRunning() && lastTimeCheck.elapsed() < MAX_SIZE_CHECK)
+    {
+        return cachedFramesWritten;
+    }
+
     QStringList strlist( QString("QUERY_RECORDER %1").arg(recordernum));
     strlist << "GET_FRAMES_WRITTEN";
 
     if (!SendReceiveStringList(strlist, 1))
     {
         LOG(VB_GENERAL, LOG_ERR, LOC + "GetFramesWritten() -- network error");
-        return -1;
+    }
+    else
+    {
+        cachedFramesWritten = strlist[0].toLongLong();
+        lastTimeCheck.restart();
     }
 
-    cachedFramesWritten = strlist[0].toLongLong();
     return cachedFramesWritten;
 }
 
diff --git a/mythtv/libs/libmythtv/remoteencoder.h b/mythtv/libs/libmythtv/remoteencoder.h
index a821a8f..2e61512 100644
--- a/mythtv/libs/libmythtv/remoteencoder.h
+++ b/mythtv/libs/libmythtv/remoteencoder.h
@@ -13,6 +13,7 @@
 #include "tv.h"
 #include "mythtypes.h"
 #include "programtypes.h"
+#include "mythtimer.h"
 
 class QStringList;
 class ProgramInfo;
@@ -93,6 +94,7 @@ class MTV_PUBLIC RemoteEncoder
     bool backendError;
     long long cachedFramesWritten;
     QMap<QString,uint> cachedTimeout;
+    MythTimer lastTimeCheck;
 };
 
 #endif
-- 
1.7.10.2

